### PR TITLE
Add method to run the compressor starting from a particular compressor-state

### PR DESCRIPTION
**Pull request by Azrenbeth**
BUILDS ON azren/config_integration_tests

Added the method continue_run() to lib.rs: 
- allows you to give a specific compressor state to start from
- runs the compressor on a chunk of the room state
- returns the compressor state and what state it got up to in the room 

This can be used to run the compressor in small chunks on rooms and still get the same state_group structure as if the room had been compressed all in one go.

### CODE REVIEW

File: src/database.rs [hunk 1]
```diff
@@ -53,12 +82,16 @@ pub fn get_data_from_db(
...
+/// * 'level_info'          -   The maximum size, current length and current head for each
+///                             level (as it was when the compressor last finished for this
+///                             room)
+pub fn reload_data_from_db(
     db_url: &str,
     room_id: &str,
     min_state_group: Option<i64>,
     groups_to_compress: Option<i64>,
     max_state_group: Option<i64>,
+    level_info: &[(usize, usize, Option<i64>)],
```
Comment (erikjohnston) [inline src/database.rs:94]: Can we not use `Level` here?

File: src/database.rs [hunk 2]
```diff
@@ -58,19 +101,126 @@ pub fn get_data_from_db(
...
+    // obtain all of the heads that aren't None from level_info
+    let level_heads: Vec<i64> = level_info
+        .iter()
+        .map(|(_, _, head)| {
+            if let Some(c) = *head {
+                return c;
+            }
+            -1
+        })
+        .filter(|c| *c > 0)
```
Comment (erikjohnston) [inline src/database.rs:140]:
```suggestion
        .filter_map(|(_, _, head)| {
            *head
        })
```

File: src/lib.rs [hunk 3]
```diff
@@ -506,6 +511,91 @@ fn output_sql(
     pb.finish();
 }
 
+/// Information about what compressor did to chunk that it was ran on
+pub struct ChunkStats {
+    // The state of each of the levels of the compressor when it stopped
+    new_level_info: Vec<(usize, usize, Option<i64>)>,
```
Comment (erikjohnston) [inline src/lib.rs]: Can't we use `Level` here?

File: src/lib.rs [hunk 4]
```diff
@@ -506,6 +511,91 @@ fn output_sql(
...
+    room_id: &str,
+    level_info: &[(usize, usize, Option<i64>)],
+) -> ChunkStats {
+    // First we need to get the current state groups
+    let (state_group_map, max_group_found) =
+        database::reload_data_from_db(db_url, room_id, Some(start), Some(chunk_size), level_info);
+
+    let original_num_rows = state_group_map
+        .iter()
+        .fold(0, |acc, (_, v)| acc + v.state_map.len());
```
Comment (erikjohnston) [inline src/lib.rs:542]:
```suggestion
        .map(|v| v.state_map.len())
        .sum();
```

File: src/compressor.rs [hunk 5]
```diff
@@ -57,6 +57,15 @@ impl Level {
...
     }
 
+    /// Creates a new level from stored state
+    pub fn restore(max_length: usize, current_chain_length: usize, current: Option<i64>) -> Level {
+        Level {
+            max_length,
+            current_chain_length,
+            current,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Azrenbeth) [reply src/compressor.rs:67]: That block is in `impl Level` - are you saying rename it to something like `new_from_save`?

[the remark this answers is not in the document]
Comment (erikjohnston) [reply src/compressor.rs:67]: Oh, I misread, nevermind!

File: src/database.rs (hunk 1 above)
Comment (Azrenbeth) [reply src/database.rs]: Done

File: src/lib.rs (hunk 3 above)
Comment (Azrenbeth) [reply src/lib.rs]: Done

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**